commit dd6fc8d8e05edb2ba7e3b9facf32313bba06065d
Author: Kalesh Singh <kaleshsingh@google.com>
Date:   Sun Oct 13 14:20:58 2024 -0700

    ltp: security: dirtyc0w_shmem: Fix backing file size
    
    Android emulates a userspace 16KiB page size on x86_64 emulators.
    The kernel however remains 4KiB and page faults are therefore handled
    in 4KiB granulatiry.
    
    A side effect of this is that there is a discrepancy in filemap faults
    which usually allows faulting off the end of a file until the next page
    boundary. In the userspace this page boundary is until the next 16KiB
    aligned address, however for the kernel the next page boudnary is until
    the next 4KiB aligned address. To solve this issue an anonymous mapping
    is inserted by the emulator's kernel to cover the range between the
    next 4KiB boundary after the file end and the next 16KiB page boundary.
    This causes the uffd register operation to fail for minor faults since
    the backing VMA is no longer shmem but anonymous.
    
    Work around this by making the backing shmem file just large enough to
    partially cover the last 4KiB page before the next 16KiB boundary after
    the file's end.
    
    Bug: 366098040
    Test: atest vts_ltp_test_x86_64:cve.cve-2022-2590_64bit#cve.cve-2022-2590_64bit
    Change-Id: Ie6e9c77870366a34b8304d87b71af71abb7f4688
    Signed-off-by: Kalesh Singh <kaleshsingh@google.com>

diff --git a/include/pgsize_helpers.h b/include/pgsize_helpers.h
index 6850f5871..a66448a4f 100644
--- a/include/pgsize_helpers.h
+++ b/include/pgsize_helpers.h
@@ -7,6 +7,8 @@
 #ifndef PGSIZE_HELPER_H
 #define PGSIZE_HELPER_H
 
+#include <stdlib.h>
+#include <string.h>
 #include <unistd.h>
 
 #define MAX_PAGE_SIZE (64*1024)
@@ -73,4 +75,28 @@ do {
         }                                                                                   \
     }                                                                                       \
 } while (0)
+
+/*
+ * Make the backing file large enough to cover the last corresponding kernel page.
+ *
+ * This is an artifact of x86_64 page size emulation on Android, to handle
+ * file_map_fault's, which does allow access to the partial page after the end
+ * of the file.
+ */
+#define SAFE_FILE_PRINTF_PGSIZE_EMULATION(file, str)                            \
+do {                                                                            \
+    if (kernel_page_size() == getpagesize()) {                                  \
+        SAFE_FILE_PRINTF(file, str);                                            \
+    } else {                                                                    \
+        int str_len = strlen(str);                                              \
+        int nr_writes = ((kernel_page_size()                                    \
+                            * (nr_pgs_to_nr_kernel_pgs(1) - 1)) / str_len) + 1; \
+        int total_len = str_len * nr_writes;                                    \
+        char *buffer = SAFE_MALLOC(total_len + 1);                              \
+        for (int i = 0; i < nr_writes; i++)                                     \
+            strcat(buffer, str);                                                \
+        SAFE_FILE_PRINTF(file, buffer);                                         \
+        free(buffer);                                                           \
+    }                                                                           \
+} while (0)
 #endif /* PGSIZE_HELPER_H */
diff --git a/testcases/kernel/security/dirtyc0w_shmem/dirtyc0w_shmem.c b/testcases/kernel/security/dirtyc0w_shmem/dirtyc0w_shmem.c
index f3d6e6423..82f2399e6 100644
--- a/testcases/kernel/security/dirtyc0w_shmem/dirtyc0w_shmem.c
+++ b/testcases/kernel/security/dirtyc0w_shmem/dirtyc0w_shmem.c
@@ -22,6 +22,7 @@
 #include <pwd.h>
 
 #include "tst_test.h"
+#include "pgsize_helpers.h"
 
 #define TMP_DIR "tmp_dirtyc0w_shmem"
 #define TEST_FILE TMP_DIR"/testfile"
@@ -62,7 +63,7 @@ static void dirtyc0w_shmem_test(void)
 	int pid;
 	char c;
 
-	SAFE_FILE_PRINTF(TEST_FILE, STR);
+	SAFE_FILE_PRINTF_PGSIZE_EMULATION(TEST_FILE, STR);
 	SAFE_CHMOD(TEST_FILE, 0444);
 
 	pid = SAFE_FORK();
